scipy/stats/morestats.py

```diff
@@ -2966,6 +2966,11 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
 
     n_zero = np.sum(d == 0)
     if n_zero > 0 and mode == "exact":
+        mode = "approx"
+        warnings.warn("Exact p-value calculation does not work if there are "
+                      "zeros. Switching to normal approximation.")
+
+    if np.unique(abs(d)).size < d.size and mode == "exact":
```

Comment:
Maybe obvious, but just a quick note that these two conditions are never true together in our test suite at the moment (see below from `codecov`). That doesn't mean changes are needed, but maybe useful to know for a `stats` reviewer when they take a look.

![image](https://user-images.githubusercontent.com/7903078/111248955-88d0a900-85d0-11eb-8a1d-d1141bec7f9f.png)